src/box/iproto.cc

```diff
@@ -2950,23 +2960,32 @@ iproto_send_listen_msg(struct evio_service *binary)
 int
 iproto_listen(const char *uri)
 {
-	if (iproto_send_stop_msg() != 0)
-		return -1;
-	evio_service_stop(&tx_binary);
-	if (uri == NULL) {
-		tx_binary.addr_len = 0;
-		return 0;
-	}
+	struct evio_service tmp;
+	memset(&tmp, 0, sizeof(tmp));
 	/*
 	 * Please note, we bind socket in main thread, and then
 	 * listen this socket in all iproto threads! With this
 	 * implementation, we rely on the Linux kernel to distribute
 	 * incoming connections across iproto threads.
 	 */
-	if (evio_service_bind(&tx_binary, uri) != 0)
+	if (uri != NULL && evio_service_bind(&tmp, uri) != 0)
 		return -1;
```

Comment:
The problem with bind + close was and still remains that the resulting IP:port can match and the new bind will fail. With close + bind it worked. So in your patch this starts to fail:
```
tarantool> box.cfg{listen = 'localhost:3313'}
---
...
tarantool> box.cfg{listen = '[::1]:3313'}
---
- error: ': failed to bind, called on fd -1'
...
```
But I don't know which behaviour is better. Close + bind and make the example above work, or bind + close but fail to rebind to the same IP:port. Perhaps you could check if the new IP:port matches the old one after you resolve the new URI and if they match - do nothing.

Reply:
Yes, here we have, error, but previously listen address is still in use, so conn = net_box.connect("3313") is work.
So problem is only in confusing message. Maybe @locker has some idea about it?

Reply:
I fixed this i stop listen, try to bind and  then listen if it's fails i go back to previous uri